sopel/modules/announce.py

```diff
@@ -34,7 +34,7 @@ def announce(bot, trigger):
         bot.reply('Announce what? I need a message to say.')
         return
 
-    size = bot.isupport.get('TARGMAX', {}).get('PRIVMSG', 1)
+    size = dict(bot.isupport.get('TARGMAX') or []).get('PRIVMSG', 1)
```

Comment:
If `TARGMAX` isn't advertised, you're going to run `[].get('PRIVMSG', 1)`?

And why not use the `default` kwarg you were so careful to add on `ISupport.get()`?

Reply:
In its raw form, `TARMAX` is parse as a list of items, that's why I transform it into a `dict( get targmax )` before accessing its `PRIVMSG` key (`dict(...).get(...)`).

And I don't like relying on default, but in that case I think I was too cautious, it could be this instead:

```suggestion
    size = dict(bot.isupport.get('TARGMAX', [])).get('PRIVMSG', 1)
```

And I need to change [this line](https://github.com/sopel-irc/sopel/blob/96c55aff852bf40bca56de49b2bc30378bf1c819/sopel/irc/isupport.py#L126) so the default value for TARGMAX is actually a tuple instead of `None`.

Reply:
> In its raw form, `TARGMAX` is parsed as a tuple of items, that's why I transform it into a `dict( get targmax )` before accessing its `PRIVMSG` key (`dict(...).get(...)`).

I still have a confuse, because [Sopel's unstable docs say](https://unstable.docs.sopel.chat/irc.html#sopel.irc.isupport.ISupport.TARGMAX) it's already a dict:

> _property_ **TARGMAX**
>
> Expose `TARGMAX` as a dict, if advertised by the server.
>
> This exposes information about the maximum number of arguments for each command:
>
> ```repl
> >>> isupport.TARGMAX
> {
>     'JOIN': None,
>     'PRIVMSG': 3,
>     'WHOIS': 1,
> }
> ```
>
> This attribute is not available if the server does not provide the right information, and accessing it will raise an `AttributeError`.

_Note: Stable docs also say this. I just quoted the unstable ones to make sure the language was as up-to-date as possible._

Now I notice that accessing stuff via property/`bot.isupport.TARGMAX` gives you _different data_ (a dict) vs. treating `bot.isupport` as a dict-ish object via `__getitem__`/`bot.isupport['TARGMAX']` (returns the internal data structure, a tuple of tuples)… and I'm mad that I didn't notice before because this is _not good_ API design IMO. Especially because the difference isn't documented, but it's just unintuitive anyway.

For _this_ PR, the solution is likely something like this:

```py
size = 1

try:
    size = bot.isupport.TARGMAX.get('PRIVMSG', 1)
except AttributeError:
    # just don't try to group targets
    pass
```

For later, the `ISupport` stuff needs revisiting. It's _probably_ not important enough for 7.1, and any changes to how the `__getitem__` access works would be "breaking" and therefore not shippable until 8.0 anyway. But I'm going to make damn sure it's on our radar. I can think of no good reason to expose the internal data structure of any `ISupport` property if it's any more complicated than a scalar.

> And I need to change [this line](https://github.com/sopel-irc/sopel/blob/96c55aff852bf40bca56de49b2bc30378bf1c819/sopel/irc/isupport.py#L126) so the default value for TARGMAX is actually a tuple instead of `None`.

Might want to change [this line](https://github.com/sopel-irc/sopel/blob/96c55aff852bf40bca56de49b2bc30378bf1c819/sopel/irc/isupport.py#L382) too, so it'll use an empty tuple instead of an empty list? That's nitpicking, yes.

Reply:
> because this is not good API design IMO

I tend to disagree with your opinion on that. By using the modern.ircdocs.horse I followed their documentation: when an item is a list, then it is parsed as a tuple (which is just an immutable list).

Although there is no documentation that states that `isupport['TARGMAX']` is a tuple, there is no documentation either that says it's a dict. There is an example given in the ISupport's docstring:

    >>> isupport = ISupport(chanlimit=(('&', None), ('#', 70)))
    >>> isupport['CHANLIMIT']
    (('&', None) ('#', 70))
    >>> isupport.CHANLIMIT  # some parameters are also properties
    {
        '&': None,
        '#': 70,
    }

And it clearly shows the difference between internal keys, and properties. It's even more clear when you look at the `CHANMODES` properties, which only return the 4 first item in the tuple, because everything outside of A, B, C, and D chan modes are unspecified.

I admit, there is a lot to document here. I think the best option is to document the `ISUPPORT_PARSERS` mapping, which contains how each value is parsed.

> so it'll use an empty tuple instead of an empty list?

No, it's just a shortcut to create an empty dict if necessary. You really tend to forget that there is a `dict(...)` surrounding stuff the way I forget about S everywhere. :grin:

Reply:
> No, it's just a shortcut to create an empty dict if necessary. You really tend to forget that there is a `dict(...)` surrounding stuff the way I forget about S everywhere. 😁

Well we all have to have _some_ quirk. 😁

Let's leave things as they are and talk about documentation improvements for 8.0, sound good? If we come up with justifiable API changes during that process, so be it.